src/libaducpal/src/ftw.c

```diff
@@ -165,318 +78,124 @@ static int construct_typeflag(const DWORD attrib, int flags)
-			 * Skip '.' and '..', because they are links to
-			 * the current and parent directories, so they
-			 * are not subdirectories.
-			 *
-			 * Skip any file that is not a directory.
-			 *
-			 * Skip all reparse points, because they might
-			 * be symbolic links. They might form a cycle,
-			 * with a directory inside itself.
-			 */
-            if (wcscmp(fn, L".") == 0 || wcscmp(fn, L"..") == 0 || (attr & FILE_ATTRIBUTE_DIRECTORY) == 0
-                || (attr & FILE_ATTRIBUTE_REPARSE_POINT))
-                continue;
-
-            ndir = malloc(sizeof *ndir);
-            if (ndir == NULL)
-            {
-                return -1;
-            }
+            ret = do_callback(func, filepath, flags, &ftwbuf);
+
+            free(filepath);
 
-            /*
-			 * Allocate space for path + backslash +
-			 *     fn + backslash + pattern + \0.
-			 */
-            fnlen = wcslen(fn);
-            buflen = dir->slashlen + fnlen + patternlen + 2;
-            buf = calloc(buflen, sizeof buf[0]);
-            if (buf == NULL)
+            if (ret != 0)
             {
-                return -1;
+                return ret;
             }
-
-            /* Copy path + backslash + fn + \0. */
-            wmemcpy(buf, dir->path, dir->slashlen);
-            wmemcpy(buf + dir->slashlen, fn, fnlen + 1);
-
-            /* Push dir to list. Enter dir. */
-            ndir->path = buf;
-            ndir->pathlen = dir->slashlen + fnlen;
-            ndir->slashlen = ndir->pathlen + 1;
-            ndir->ffh = INVALID_HANDLE_VALUE;
-            ndir->next = dir;
-            dir = ndir;
-            goto loop; /* Continue outer loop. */
-        } while (FindNextFileW(dir->ffh, &dir->ffd) != 0);
-    closeffh:
-        if (GetLastError() != ERROR_NO_MORE_FILES)
-        {
-            dir->path[dir->pathlen] = '\0';
-            // oops(dir->path);
-            rval = 1;
         }
-        FindClose(dir->ffh);
-
-    popdir:
-        /* Pop dir from list, free dir, but never free dir0. */
-        free(dir->path);
-        if ((ndir = dir->next))
-            free(dir);
-        dir = ndir;
     }
 
-    return rval;
-}
-
-static int _wnftw(const wchar_t* wpath, NFTW_FUNC_T func, int fd_limit, int flags)
-{
-    if (fd_limit <= 0)
+    if (level == 0)
     {
-        return 0;
+        // Do a final callback for the folder that was initially passed in
+        ret = do_callback(func, path, flags, &ftwbuf);
     }
 
-    const size_t l = wcslen(wpath);
-    if (l > PATH_MAX)
+    if (dp != NULL)
     {
-        errno = ENAMETOOLONG;
-        return -1;
+        ADUCPAL_closedir(dp);
     }
 
-    wchar_t pathbuf[PATH_MAX + 1];
-    wcscpy_s(pathbuf, sizeof(pathbuf) / sizeof(pathbuf[0]), wpath);
-
-    make_forward_slash(pathbuf);
-
-    return do_wnftw(pathbuf, func, fd_limit, flags, NULL);
+    return ret;
 }
 
+// Returns 0 on success, and -1 if an error occurs.
 int ADUCPAL_nftw(const char* path, NFTW_FUNC_T func, int nopenfd, int flags)
```

Comment:
there's findFilesInDir in file_utils that uses DIRENT, so it could be tweaked/ported to C to allow returning "DIR" file types as well as regular files and a C wrapper on top of the cpp module could be added to adapt it.

Or better yet, you could break away from nftw semantics and create a "foreach_file" API that just calls the passed-in callback after stat'ing each file that would include all types of files -- directory, regular files, symlinks, fifo pipes, etc.

Reply:
findFilesInDir doesn't handle subdirs, so that doesn't add any additional value here.  opendir() is the "standard" way of doing this, which is what im using